lib/wallaroo/core/sink/sink_phase.pony

```diff
@@ -36,13 +39,18 @@ trait SinkPhase
   =>
     _invalid_call(__loc.method_name()); Fail()
 
-  fun ref receive_barrier(input_id: RoutingId, producer: Producer,
-    barrier_token: BarrierToken)
+  fun ref run[D: Any val](metric_name: String,
+    pipeline_time_spent: U64, data: D, key: Key, event_ts: U64,
+    watermark_ts: U64, i_producer_id: RoutingId, i_producer: Producer,
+    msg_uid: MsgId, frac_ids: FractionalMessageId, i_seq_id: SeqId,
+    latest_ts: U64, metrics_id: U16, worker_ingress_ts: U64)
   =>
     _invalid_call(__loc.method_name()); Fail()
 
-  fun ref early_prepare_for_rollback() =>
-    None
+  fun ref receive_barrier(input_id: RoutingId, producer: Producer,
```

Comment:
What does the `Bool` represent?

Reply:
The `Bool` return value represents whether or not the barrier is complete and is used in cases where `BarrierSinkPhase`'s `completion_notifies_sink` flag is false.

Apropos, it's quite likely that `NormalSinkPhase.receive_barrier()` at line 144 is buggy: i.e., in the case where the new phase has only 1 blocking input/token expected.

Reply:
Can you specify in a comment?

Reply:
I added a docstring to the trait's definition.

Reply:
Isn't this the trait's definition? I don't see the docstring